folio-org/mod-circulation
# CIRC-716: Refund lost item fees on override renewal.

https://issues.folio.org/browse/CIRC-716 - BE - Declared lost: renewal (effect on lost item fees) - SET COST

## Purpose
Refund lost item fees when `Declared lost` item is renewed through override. 

## Implementation notes

The main business logic is the same as for check-in (https://issues.folio.org/browse/CIRC-717). The main changes is to integrate the refund logic with override renewal. For this, new  [`RenewalFeeProcessingStrategy`](https://github.com/folio-org/mod-circulation/pull/579/files#diff-f9787b2ed191b119232110fc10e35d2f) class has been introduced with two implementations:
* [`RegularRenewalFeeProcessingStrategy`](https://github.com/folio-org/mod-circulation/pull/579/files#diff-c03a04597161f4b24604313d8bab425a) - charges overdue fines only, if any. Applies to renewal by barcode/by id operations;
*  [`OverrideRenewalFeeProcessingStrategy`](https://github.com/folio-org/mod-circulation/pull/579/files#diff-ec88929ebe68a4f1265910900aa0a793) - refunds lost item fees, if any and then executes the `RegularRenewalFeeProcessingStrategy` to do the usual fee processing staff. Applies only to override renewal operation;

## Approach
### Production code:
* Add new optional [`overrideServicePointId`](https://github.com/folio-org/mod-circulation/pull/579/files#diff-da6e4435f06c2a4f0886defa03e5b77d) property for `override-renewal-by-barcode-request` request. Bump up circulation interface version;
* Add the new fee processing strategies for renew and override - [renew](https://github.com/folio-org/mod-circulation/pull/579/files#diff-c03a04597161f4b24604313d8bab425a), [override](https://github.com/folio-org/mod-circulation/pull/579/files#diff-ec88929ebe68a4f1265910900aa0a793);
* Assign overdue fines for lost items when returned - [OverdueFineCalculatorService.java](https://github.com/folio-org/mod-circulation/pull/579/files#diff-f30d04d1dcd201545e5c2adfce52f4a9);
* Add small adapter method to refund lost fees on renewal - [LostItemFeeRefundService.java](https://github.com/folio-org/mod-circulation/pull/579/files#diff-2403ad4ea13245175169b2f8de582d52)

### Test code:
* Reuse all existing tests for refund lost item fees on check-in by moving them into an [abstract ](https://github.com/folio-org/mod-circulation/pull/579/files#diff-88d86c58f33eee5a890956333e5c5f97) class and extending by [OverrideRenewDeclaredLostItemTest.class](https://github.com/folio-org/mod-circulation/pull/579/files#diff-f54339b62c01186cd0bc05ff0590e60a) and [CheckInDeclaredLostItemTest.class](https://github.com/folio-org/mod-circulation/pull/579/files#diff-0a790d253e0c77c940e37b7c33983de1) (the existing one);
* Introduce spring testing/context/bean test dependencies;
* Refactor [RestAssuredClient.java](https://github.com/folio-org/mod-circulation/pull/579/files#diff-a91fbb2b080873d7a51699aacbf989b8) to allow accepting an object as body representation and add [JsonObjectJacksonSerializer.class](https://github.com/folio-org/mod-circulation/pull/579/files#diff-412312e74bcb98c07537ad51b4483379) to allow proper jackson serialization of `JsonObject` class.
* Introduce new parameterized [ResourceClient](https://github.com/folio-org/mod-circulation/pull/579/files#diff-f1987ccf077915692c860fd96ca5271d);
* Add [OverrideRenewalFixture.class](https://github.com/folio-org/mod-circulation/pull/579/files#diff-69ac5f987c36736b84e18d3af6aeffb5);
* Add [spring context configuration for tests](https://github.com/folio-org/mod-circulation/pull/579/files#diff-99e156d9d8e8e6689b8e53a6019b2923)


#### TODOS and Open Questions


## Learning


## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.
